### PR TITLE
initial commit of generic refactor

### PR DESCRIPTION
**Pull request by groovecoder**
Should show codesy widget on:
- GitHub issues
- BitBucket issues
- SourceForge tickets

### CODE REVIEW

File: data/launch.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          console.log("getting codesyImgUrl");
+          var codesyImgUrl = self.options.codesyImgUrl;
+          console.log("codesyImgUrl: "+codesyImgUrl);
+          codesy.options.domain = self.options.codesyDomain;
+          console.log("domain: "+codesy.options.domain);
+
+          mission.before_append && mission.before_append (self.options);
+          
+
+          codesy.appendForm(mission, codesyImgUrl, codesyToken)
```
Comment (groovecoder) [inline data/launch.js:25]: In particular, I need @jdungan to help me figure out why this appendForm call triggers an error:

```
A promise chain failed to handle a rejection.
```

Even though I added a `.fail()` function.

Comment (jdungan) [inline data/launch.js:25]: You've added a console.log line in codesy.appendForm before creating the deferred object.  Is this where it fails?

Comment (groovecoder) [inline data/launch.js:25]: Removing it didn't change anything. I'm leaning towards taking these deferred/promises out. I think they may not work with the way Firefox content scripts communicate with other js modules in the add-on.

Comment (jdungan) [inline data/launch.js:25]: $,ajax is a deferred object so you can't avoid it. Why do you merge it and let me take a crack at it. Or let's get together and walk through the code.

Comment (groovecoder) [inline data/launch.js:25]: What are you doing later tonight? :)

**Comment by groovecoder** [comment]
Unbelievable. One errant semi-colon. https://stackoverflow.com/questions/21359206/a-promise-chain-failed-to-handle-a-rejection-while-loading-add-on

**Comment by groovecoder** [comment]
Whew, this is finally ready for a spot-check by someone else. @mwic? @jgmize?

**Pull request merged**